Implement admin control panel for dynamic feature gating

MegrezZhu opened the pull request:
## Overview

[Design Doc of Dynamic Feature Gating](https://docs.google.com/document/d/1FjbG2Cb0OnDVuE36jzux2Sn0sybMQozkk3hjkfeMoC0/edit?usp=sharing)

1. This PR fixes or fixes part of N/A.
2. This PR does the following:
    * added a new tab `feature` at the admin page
    * added feature flag results in the `admin-data.service.ts`

Screenshot for the admin panel UI:

![image](https://user-images.githubusercontent.com/21364971/90851189-ee61fd00-e341-11ea-9871-290dd2c3533d.png)

![image](https://user-images.githubusercontent.com/21364971/90851125-c4103f80-e341-11ea-9cad-601eb60e2cc0.png)



/cc @BenHenning @vinitamurthi

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [ ] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

MegrezZhu committed:
Add storage models

MegrezZhu committed:
Merge branch 'develop' into dynamic-feature-gating-backend-dev

MegrezZhu committed:
Rename models, add validation jobs and change domain object

MegrezZhu committed:
Merge branch 'develop' into dynamic-feature-gating-backend-dev

MegrezZhu committed:
Fix method naming

MegrezZhu committed:
Fix import error

MegrezZhu committed:
Fix test issues

MegrezZhu committed:
Improve coverage, fix linting errors, add codeowner

MegrezZhu committed:
Remove debug statements

MegrezZhu committed:
Add code owner entries

MegrezZhu committed:
Change validator id regex

MegrezZhu committed:
Add domain objects

MegrezZhu committed:
Add docstrings

MegrezZhu committed:
Improve test coverage

MegrezZhu committed:
Added requested changes

MegrezZhu committed:
Fix test errors

MegrezZhu committed:
Merge branch 'storage-dev' into domain-dev

MegrezZhu committed:
Fix test errors

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Change code owners

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Add missing docstrings, fix test errors

MegrezZhu committed:
Add service methods

MegrezZhu committed:
Add docstrings

MegrezZhu committed:
Improve test

MegrezZhu committed:
Add schema_version, improve docstrings

MegrezZhu committed:
Merge branch 'develop' into storage-dev

MegrezZhu committed:
Merge branch 'storage-dev' into domain-dev

MegrezZhu committed:
Styling

MegrezZhu committed:
Fix audit job tests

MegrezZhu committed:
Merge branch 'storage-dev' into domain-dev

MegrezZhu committed:
Add schema version

MegrezZhu committed:
Merge branch 'domain-dev' into service-dev

MegrezZhu committed:
Add method, fix test errors

MegrezZhu committed:
Add necessary imports

MegrezZhu committed:
Styling, fix codeowners

MegrezZhu committed:
Styling

MegrezZhu committed:
Add controllers

MegrezZhu committed:
Merge branch 'storage-dev' into domain-dev

MegrezZhu committed:
Add method for feature, add dummy instance for demo

MegrezZhu committed:
Merge branch 'domain-dev' into controller-dev

MegrezZhu committed:
Add test

MegrezZhu committed:
Merge branch 'develop' into domain-dev

MegrezZhu committed:
Merge branch 'domain-dev' into service-dev

MegrezZhu committed:
Merge branch 'service-dev' into controller-dev

MegrezZhu committed:
Fix linting errors on keyword args

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Refactor filter to use (op, value) schema

MegrezZhu committed:
Add requested change

MegrezZhu committed:
Add browser type check

MegrezZhu committed:
Change docstring

MegrezZhu committed:
Add rquested changes

MegrezZhu committed:
Change test class name

MegrezZhu committed:
Move Registry

MegrezZhu committed:
Add front end domain objects

MegrezZhu committed:
Add front end domain objects

MegrezZhu committed:
Add code owners

MegrezZhu committed:
Add code owners

MegrezZhu committed:
Merge branch 'develop' into fe-domain-dev

MegrezZhu committed:
Fix linting error, add result object

MegrezZhu committed:
Fix linting error, add result object

MegrezZhu committed:
Remove debugging statements

MegrezZhu committed:
Remove debugging statements

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Improve test case names

MegrezZhu committed:
Fix docstring

MegrezZhu committed:
Fix test error

MegrezZhu committed:
Rename result object

MegrezZhu committed:
Rename result object

MegrezZhu committed:
Add client context object, fix test error

MegrezZhu committed:
Add client context object, fix test error

MegrezZhu committed:
Improve test naming, add requested changes

MegrezZhu committed:
Rename helper method

MegrezZhu committed:
Remove metadata object, add comment for tests

MegrezZhu committed:
Add app_version_flavor filter

MegrezZhu committed:
Change none value handling

MegrezZhu committed:
Merge branch 'domain-dev' into service-dev

MegrezZhu committed:
Sync interface, refactor and improve test

MegrezZhu committed:
Edit codeowners

MegrezZhu committed:
Fix linter error

MegrezZhu committed:
Merge branch 'service-dev' into controller-dev

MegrezZhu committed:
Sync interfaces, fix codeowner

MegrezZhu committed:
Fix docstrings

MegrezZhu committed:
Merge branch 'service-dev' into controller-dev

MegrezZhu committed:
Merge branch 'controller-dev' into fe-domain-dev

MegrezZhu committed:
Remove metadata object

MegrezZhu committed:
Remove metadata object

MegrezZhu committed:
Merge branch 'develop' into domain-dev

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Add comparison support to flavor filters

MegrezZhu committed:
Merge branch 'domain-dev' into service-dev

MegrezZhu committed:
Merge branch 'service-dev' into controller-dev

MegrezZhu committed:
Merge branch 'controller-dev' into fe-domain-dev

MegrezZhu committed:
Fix bug with None version

MegrezZhu committed:
Merge branch 'domain' into service-dev

MegrezZhu committed:
Change expected payload format

MegrezZhu committed:
Merge branch 'controller' into fe-domain-dev

MegrezZhu committed:
Add frontend services

MegrezZhu committed:
Improve test and docstrings

MegrezZhu committed:
Fix import sorting

MegrezZhu committed:
Add test, remove fdescribe

MegrezZhu committed:
Fix flavor ordering, fix docstring

MegrezZhu committed:
Add requested change

MegrezZhu committed:
Merge branch 'domain-dev' into service-dev

MegrezZhu committed:
Sync interface

MegrezZhu committed:
Fix redirect bug, add codeowner, fix fileoverview

MegrezZhu committed:
Merge branch 'service-dev' into controller-dev

MegrezZhu committed:
Merge branch 'controller-dev' into fe-domain-dev

MegrezZhu committed:
Merge branch 'fe-domain' into fe-service-dev

MegrezZhu committed:
Merge branch 'develop' into service-dev

MegrezZhu committed:
Add features tab

MegrezZhu committed:
Merge branch 'develop' into service-dev

MegrezZhu committed:
Merge branch 'develop' into controller

MegrezZhu committed:
Rename files, improve comment & test

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Add interface description

MegrezZhu committed:
Link todo with issue

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Delete validation

MegrezZhu committed:
Delete validation

MegrezZhu committed:
stash

MegrezZhu committed:
Make properties static

MegrezZhu committed:
Add guard for initialization, add docstring

MegrezZhu committed:
Fix test

MegrezZhu committed:
Sync change

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Improve interface

MegrezZhu committed:
Add static UI

MegrezZhu committed:
Rename files using new pattern

MegrezZhu committed:
Rename FeatureFlagResults -> FeatureStatusSummary

MegrezZhu committed:
Merge branch 'fe-domain-dev' into fe-admin-dev

MegrezZhu committed:
Add interaction

MegrezZhu committed:
Rename files & fix typo

MegrezZhu committed:
Fix spacing, improve docstring

MegrezZhu committed:
Merge branch 'service-dev' into controller-dev

MegrezZhu committed:
Rename files

MegrezZhu committed:
Sync interface

MegrezZhu committed:
Sync renaming

MegrezZhu committed:
Fix naming

MegrezZhu committed:
Rename files

MegrezZhu committed:
Fix naming

MegrezZhu committed:
Add validation and moving

MegrezZhu committed:
Merge branch 'develop' into service-dev

MegrezZhu committed:
Merge branch 'develop' into fe-domain-new-dev

MegrezZhu committed:
Add test & codeowners

MegrezZhu committed:
Merge branch 'service-dev' into controller-dev

MegrezZhu committed:
Improve error handling & add test for invalid rule update

MegrezZhu committed:
Merge branch 'controller-dev' into controller

MegrezZhu committed:
Merge branch 'controller-dev' into fe-service-dev

MegrezZhu committed:
Merge branch 'controller-dev' into fe-service-dev

MegrezZhu committed:
Rename files, sync changes

MegrezZhu committed:
Fix linting error

MegrezZhu committed:
Merge branch 'controller-dev' into fe-service-dev

MegrezZhu committed:
Improve interface

MegrezZhu committed:
Merge branch 'fe-service-dev' into fe-admin-dev

MegrezZhu committed:
Fix type, fix codeowner, add docstring

MegrezZhu committed:
Sync interface

MegrezZhu committed:
Merge branch 'develop' into service-dev

MegrezZhu committed:
Merge branch 'service-dev' into controller-dev

MegrezZhu committed:
Merge branch 'controller-dev' into fe-service-dev

MegrezZhu committed:
Merge branch 'fe-service-dev' into fe-admin-dev

MegrezZhu committed:
Remove debug statements

MegrezZhu committed:
Merge branch 'fe-service-dev' into fe-admin-dev

MegrezZhu committed:
Sync interface changes, fix test

MegrezZhu committed:
Fix request bug

MegrezZhu committed:
Merge branch 'fe-service-dev' into fe-admin-dev

MegrezZhu committed:
Merge branch 'develop' into fe-service-dev

MegrezZhu committed:
Merge branch 'fe-service-dev' into fe-admin-dev

MegrezZhu committed:
Add missing test, fix test error, fix linting error

MegrezZhu committed:
Merge branch 'fe-service' into fe-admin-dev

MegrezZhu committed:
Fix test errors

MegrezZhu committed:
Fix linting error, improve test coverage

MegrezZhu committed:
Merge branch 'develop' into controller-dev

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Merge branch 'controller-dev' into fe-service-dev

MegrezZhu committed:
Sync changes

MegrezZhu committed:
Merge branch 'fe-service' into fe-admin

MegrezZhu committed:
Sync change, simplify prompting

MegrezZhu committed:
Merge branch 'develop' into controller-dev

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Fix typo

MegrezZhu committed:
Merge branch 'develop' into controller-dev

MegrezZhu committed:
Fix errors after merging develop

MegrezZhu commented:
@vojtechjelinek @seanlip @vinitamurthi @nithusha21 @DubeySandeep @aks681 @bansalnitish @srijanreddy98 @jameesjohn
PTAL at this pull request as reviewers/code owners, note that this PR is chained with previous PRs, so the changed files shown in this PR includes changes that are introduced in previous PRs, please just look at the files listed in the description of this pull request for now, and also feel free to take another pass once the previous PRs get merged and the delta is clean.

seanlip commented:
@MegrezZhu perhaps you could only assign people who are codeowners of the files you want reviewed? For instance, I'm codeowner of only core/controllers/base.py and have already left comments about it in a separate PR.

vinitamurthi commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.spec.ts:
```diff
@@ -0,0 +1,374 @@
+      .and.resolveTo(null);
+  }));
+
+  beforeEach(async(() => {
+    fixture = TestBed.createComponent(AdminFeaturesTabComponent);
+    component = fixture.componentInstance;
+
+    component.ngOnInit();
+  }));
```
Can this be added to the above beforeEach function?

MegrezZhu replied:
Yes, combined the two beforeEach functions.

MegrezZhu committed:
Add requested changes

vojtechjelinek reviewed (approved):
LGTM from code owner's perspective. Thanks!

MegrezZhu commented:
De-assigned codeowners of files only changed in previous PRs.

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Merge branch 'develop' into fe-service-dev

MegrezZhu committed:
Add tests, remove frontend coverage exception

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Merge branch 'controller-dev' into fe-service-dev

MegrezZhu committed:
Combine two beforeEach

MegrezZhu committed:
Merge branch 'fe-service' into fe-admin

BenHenning commented on core/templates/pages/admin-page/admin-page.directive.html:
```diff
@@ -23,6 +23,10 @@
   </admin-config-tab>
 
+  <admin-features-tab ng-if="$ctrl.isFeaturesTabOpen()"
+                      [set-status-message]="$ctrl.setStatusMessage">
```
Why ``[set-status-message]`` vs ``set-status-message``?

MegrezZhu replied:
It's needed for passing the function from the parent angularjs component to the child angular component as per [this wiki page](https://github.com/oppia/oppia/wiki/Angular-Migration#4-changing-the-bindtocontroller).
Now it's been changed to `(set-status-message)` according to this [comment thread](https://github.com/oppia/oppia/pull/10334/files#r473693110)

BenHenning replied:
Thanks for the context!

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.constants.ts:
```diff
@@ -0,0 +1,41 @@
+  public static ALLOWED_CLIENT_TYPES = ['Android', 'Web'];
+
+  public static ALLOWED_BROWSER_TYPES = [
+    'Chrome', 'Edge', 'Safari', 'Firefox', 'Others'];
```
If we can, we should probably have these constants in a constants.ts enum to ensure consistency across both the frontend & backend for all cases when we refer to these. Ditto for client types & app versions.

MegrezZhu replied:
Done, moved client_types, browser_types, and app_version_flavors to the constants.ts, also updated backend code to use definitions in constants.ts instead.

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.constants.ts:
```diff
@@ -0,0 +1,41 @@
+  public static ALLOWED_BROWSER_TYPES = [
+    'Chrome', 'Edge', 'Safari', 'Firefox', 'Others'];
+
+  public static APP_VERSION_REGEXP = /^\d+(?:\.\d+)*$/;
```
Does this account for app version hashes, or do we specifically want to restrict version types to non-hash formats when implementing filters?

MegrezZhu replied:
This only takes the numeric part of the version literal into account, I think this is expected -- we don't want the administrator to type in hashes, and we already have another filter for flavors, so I restricted it to be non-hash format.

BenHenning replied:
Yup, makes sense. Thanks!

Suggest adding a comment to describe your intent here.

MegrezZhu replied:
Done.

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.directive.html:
```diff
@@ -0,0 +1,237 @@
+<div *ngFor="let featureFlag of featureFlags; index as featureIndex;" class="container oppia-page-card oppia-long-text">
```
Please add screenshots and/or videos demoing the new tab in action for reviewers to better understand how the new UI ties back to the code being introduced here.

MegrezZhu replied:
Ok, I will attach a screenshot to the description of this PR.

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.ts:
```diff
@@ -0,0 +1,249 @@
+          case FeatureStage.DEV:
+            return option === 'dev';
+          case FeatureStage.TEST:
+            return option !== 'prod';
```
Why not do ``option === 'test'`` and similar for prod to keep things really clear?

MegrezZhu replied:
Done.

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.ts:
```diff
@@ -0,0 +1,249 @@
+            return option !== 'prod';
+          case FeatureStage.PROD:
+            return true;
+        }
```
Should this default to false?

MegrezZhu replied:
Do you mean that the `FeatureStage.PROD` case should return false, or that there should be a default case that always return false?
If the former, the `FeatureStage.PROD` should return true because the return value of this filter function means what mode the prod feature can be enabled in, and for prod feature it will be all.

BenHenning replied:
Makes sense. I realized a bit later that this was enforcing the ordering (I hadn't caught that initially), but this makes sense. Thanks for adding the default.

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.ts:
```diff
@@ -0,0 +1,249 @@
+
+import cloneDeep from 'lodash/cloneDeep';
+
+
```
Why extra newline? Prefer to omit for consistency.

MegrezZhu replied:
Done.

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.ts:
```diff
@@ -0,0 +1,249 @@
+    .keys(PlatformParameterFilterType)
+    .map(key => PlatformParameterFilterType[key]);
+
+  readonly filterTypeToContext: {
```
Nice implementation design. This is a good use of generic programming to make things simpler & more declarative.

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.ts:
```diff
@@ -0,0 +1,249 @@
+  addNewRuleToBottom(feature: PlatformParameter): void {
+    feature.rules.push(
+      this.ruleFactory.createFromBackendDict({
+        filters: [{
+          type: PlatformParameterFilterType.ServerMode,
+          conditions: [['=', ServerMode.Dev.toString()]]
+        }],
```
Maybe save this as a private constant to represent the default filter (and to provide context that's what this is).

MegrezZhu replied:
Done.

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.ts:
```diff
@@ -0,0 +1,249 @@
+      `Update feature '${feature.name}'.`
+    );
+    if (commitMessage === null) {
+      return;
```
Suggest setting a status message in this case rather than silently failing.

MegrezZhu replied:
It's not a fail, the commitMessage is null when the user clicks the cancel button on the browser prompt, I think it's intuitive to silently abort the update.

BenHenning replied:
Ah, to be clear I was considering the case when the user gets the confirmation dialog & doesn't fill in a commit message, then clicks confirm. Won't that look the same as cancelling with the current logic?

MegrezZhu replied:
As long as the user clicks the confirm button on the dialog, the `commitMessage` is a string, which may be an empty string if the user leave the message empty. 
Given that the user clicks the confirm button, I think we should proceed the update, and we don't have a strong constraint such that the commit message cannot be empty.

BenHenning replied:
Ah, I see. Okay, sounds reasonable.

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.ts:
```diff
@@ -0,0 +1,249 @@
+      this.adminTaskManager.startTask();
+
+      await this.apiService.updateFeatureFlag(
+        feature.name, commitMessage, feature.rules);
```
In the event the update is successful, does this correctly update the backups? Or, will restoring the backups potentially override the local copies of flags already updated in the backend? I suggest adding tests for this case.

MegrezZhu replied:
Yes I missed the update... added codes & tests for that.

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.spec.ts:
```diff
@@ -0,0 +1,372 @@
+    });
+  });
+
+  describe('.addNewRuleToBottom', () => {
```
Per this, the above test, and other tests based on indexes or directionality: are these actually verifying that the correct thing is being added/removed/changed at the correct location? If not, please make sure they are. It seems like you could reverse the directionality of the prod functions for addNewRuleTo{Top,Bottom} and these tests would still pass.

MegrezZhu replied:
Improved these changes by adding more rules/filter/conditions, and checking that only the intended ones are changed and the others are untouched.

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.spec.ts:
```diff
@@ -0,0 +1,372 @@
+    });
+  });
+
+
```
Here, above, and elsewhere: please remove unnecessary extra newlines.

MegrezZhu replied:
Done.

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.spec.ts:
```diff
@@ -0,0 +1,372 @@
+      expect(featureFlag.rules.length).toBe(1);
+    });
+
+    it('should not proceed if the user does\'t confirm', () => {
```
```suggestion
    it('should not proceed if the user doesn\'t confirm', () => {
```

MegrezZhu replied:
Done.

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.spec.ts:
```diff
@@ -0,0 +1,372 @@
+      component.setStatusMessage = setStatusSpy;
+
+      adminTaskManagerService.finishTask();
+      mockConfirmResult(true);
```
For this & other tests above/below that use similar mock setups: I suggest always keeping the mock calls in relevant methods rather than defaulting it. Otherwise, it's hard to verify that the happy path tests are actually verifying the right thing since the reader needs to go searching for wherever the other part of the test arrangement is happening.

MegrezZhu replied:
Done.

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.spec.ts:
```diff
@@ -0,0 +1,372 @@
+import { WindowRef } from 'services/contextual/window-ref.service';
+
+
+describe('Admin page feature tab', function() {
```
Suggest also making sure that all of the different structural changes happening in the prod code are being verified in tests below. We definitely don't want bugs in the admin panel leading to unexpected changes to feature flags that, in the worst case, actually pass validation and potentially cause immediate effects on users.

MegrezZhu replied:
Done, improved tests for them.

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.ts:
```diff
@@ -0,0 +1,249 @@
+    featureFlag.rules = backup.rules;
+  }
+
+  onFilterTypeSelectionChanged(filter: PlatformParameterFilter): void {
```
Perhaps this should be ``removeTopFilterCondition`` or similar?

MegrezZhu replied:
This is called when the type of a filter is changed, in such cases all existsing conditions (operators, values) are not applicable any longer, so this method clear all conditions of that filter.
Renamed it to `clearFilterConditions` to make it clearer.

BenHenning reviewed (changes requested):
Thanks @MegrezZhu! I still need to try this out, but the code looks good. Most just have various nits & suggestions for improvements. As indicated in one of my comments, it would also be really helpful to see screenshots.

MegrezZhu committed:
Add checks for new feature

bansalnitish reviewed (approved):
LGTM for codeowner files!

srijanreddy98 commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.ts:
```diff
@@ -0,0 +1,249 @@
+  templateUrl: './admin-features-tab.directive.html'
+})
+export class AdminFeaturesTabComponent implements OnInit {
+  @Input() setStatusMessage: (msg: string) => void;
```
@MegrezZhu passing functions to @Input is not considered good practice in angular. The better way is to use @Output.
make sure to import EventEmitter from @angular/core, otherwise, you will run into errors.
```suggestion
  @Output() setStatusMessage = new EventEmitter<string>();
```

And wherever you are calling setStatusMessage, change it to `this.setStatusMessage.emit('the message');

In the HTML file (https://github.com/oppia/oppia/pull/10334/files?file-filters%5B%5D=.html&file-filters%5B%5D=.ts&file-filters%5B%5D=No+extension#r473447358) where you have used `[set-status-message]`, change to this  
```(set-status-message)="$ctrl.setStatusMessage($event)"```

MegrezZhu replied:
Done, thank you for the detailed instructions!

srijanreddy98 commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.ts:
```diff
@@ -0,0 +1,249 @@
+
+@Component({
+  selector: 'admin-features-tab',
+  templateUrl: './admin-features-tab.directive.html'
```
This html file should be renamed to ./admin-features-tab.component.html

MegrezZhu replied:
Done.

srijanreddy98 commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.ts:
```diff
@@ -0,0 +1,249 @@
+  PlatformParameterFilter,
+  ServerMode,
+} from 'domain/platform_feature/platform-parameter-filter-object.factory';
+
```
Optional: The component looks really big. Can you break it down in smaller dumb components and have this as a main smart component?

MegrezZhu replied:
You are right, it would be better to break it down, but I prefer to keep it this way for now and leave it as a future improvement because of the timing issue..

BenHenning replied:
@srijanreddy98 would it be acceptable to file an issue to track this, instead? Besides the timing sensitivity of getting this PR merged, the size of the PR & its position in the chain over other changes doesn't lend itself well to including a large-ish refactor of splitting this file up.

srijanreddy98 replied:
@BenHenning @MegrezZhu of course! I kept it as optional because this PR needs to go in soon. You can do that in a follow-up PR after all of the PRs' pertaining to dynamic feature gating have been merged.

srijanreddy98 reviewed (changes requested):
Thanks @MegrezZhu! I have some concerns regarding the usage of @Input. Can you take a look!

nithusha21 reviewed (approved):
Thanks @BenHenning for the thorough review. I actually have nothing to add!

DubeySandeep reviewed (approved):
LGTM for the codeowner files!

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Merge branch 'develop' into fe-service-dev

MegrezZhu committed:
Fix browser check

MegrezZhu committed:
Merge branch 'fe-service' into fe-admin-dev

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Add requested changes

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.spec.ts:
```diff
@@ -130,37 +135,52 @@ describe('Admin page feature tab', function() {
       component.moveRuleUp(featureFlag, 1);
-      expect(featureFlag.rules[0].valueWhenMatched).toBeFalse();
+
+      expect(featureFlag.rules[0].valueWhenMatched).toEqual('1');
```
Maybe add a clarifying comment, e.g.:

```python
# Verify rule order 'original' -> '1' -> '2' now reflects the moved rule '1'.
```

Ditto elsewhere for tests where the asserted behavior may not be immediately obvious when glancing at the test.

MegrezZhu replied:
Done.

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.ts:
```diff
@@ -214,12 +212,14 @@ export class AdminFeaturesTabComponent implements OnInit {
         feature.name, commitMessage, feature.rules);
 
-      this.setStatusMessage('Saved successfully.');
+      this.featureFlagNameToBackupMap.set(feature.name, cloneDeep(feature));
```
Suggest also verifying that if the update fails, the in-memory backup is not changed.

MegrezZhu replied:
Done.

BenHenning reviewed (changes requested):
Thanks @MegrezZhu! Some thoughts on the design:

- This doesn't seem to be matching the mocks from the design document. Why the deviation?
- What does 'others' mean when matching browser types?
- We should probably only enable the 'save' and 'clear changes' buttons if there are changes
- The 'No condition specified' sometimes shows up, and sometimes it's a defaulted server mode. This feels a bit inconsistent. Maybe always show 'No condition specified'?
- Shouldn't app versions match exactly against ``(\d+)\.(\d+)\.(\d+)``? At the moment, devs can add filters for '1', '1.2', '1.2.3.4', etc. which will never match.
- What do multiple features look like? Can you add a screenshot for that? Are they in separate white boxes?
- Redundant filters & conditions don't seem to trigger any validation errors--is that intended?
- If it's not too much work, I think rearranging the UI a bit might make it a bit more usable, specifically:
  - The 'Add to top'/'Add to bottom' buttons should probably say 'Add rule to top/bottom
  - The rule add to top/bottom buttons should be above and below the list of rules, respectively
  - Suggest using a darker color for all the lines--see the jobs tab for reference. The current color is a bit light & hard to see
  - Suggest left aligning the rules & have them take up the full space available in the tab; there's a lot of space being wasted on the left (moving the add rule buttons as suggested should help with this)
  - The 'delete' & 'move' buttons should be reversed & on the right side of the rules (e.g. delete is always on the outside)
  - The add buttons for filters & conditions should probably be on the bottoms of the lists because they are always adding to the bottom
  - The conditions & filters can probably also similarly be left aligned to give them more space

MegrezZhu committed:
Add requested changes & adjust UI

MegrezZhu committed:
Merge branch 'develop' into fe-admin-dev

MegrezZhu commented:
> Thanks @MegrezZhu! Some thoughts on the design:
> 
> * This doesn't seem to be matching the mocks from the design document. Why the deviation?
> * What does 'others' mean when matching browser types?
> * We should probably only enable the 'save' and 'clear changes' buttons if there are changes
> * The 'No condition specified' sometimes shows up, and sometimes it's a defaulted server mode. This feels a bit inconsistent. Maybe always show 'No condition specified'?
> * Shouldn't app versions match exactly against `(\d+)\.(\d+)\.(\d+)`? At the moment, devs can add filters for '1', '1.2', '1.2.3.4', etc. which will never match.
> * What do multiple features look like? Can you add a screenshot for that? Are they in separate white boxes?
> * Redundant filters & conditions don't seem to trigger any validation errors--is that intended?
> * If it's not too much work, I think rearranging the UI a bit might make it a bit more usable, specifically:
>   
>   * The 'Add to top'/'Add to bottom' buttons should probably say 'Add rule to top/bottom
>   * The rule add to top/bottom buttons should be above and below the list of rules, respectively
>   * Suggest using a darker color for all the lines--see the jobs tab for reference. The current color is a bit light & hard to see
>   * Suggest left aligning the rules & have them take up the full space available in the tab; there's a lot of space being wasted on the left (moving the add rule buttons as suggested should help with this)
>   * The 'delete' & 'move' buttons should be reversed & on the right side of the rules (e.g. delete is always on the outside)
>   * The add buttons for filters & conditions should probably be on the bottoms of the lists because they are always adding to the bottom
>   * The conditions & filters can probably also similarly be left aligned to give them more space

@BenHenning thanks for the reviews!

* The UI is changed to adapt to the new dimension of conditions
* `Other` is the same as the value returned by `detectBrowserType`, that stands for all other types of browsers
* Added checks for changed rules, now the 'save' and 'clear' button will be disabled when there isn't any change
* Changed default rule filters, now 'No condition specified' always shows up
* Changed regex for app version to `(\d+)(?:\.\d+){2}`
* Yes they are in separate white boxes, I added a screenshot for this case.
* Since redundant filters/conditions are not considered errors in the backend validation, I didn't add explicit checks for such cases, maybe we can add an error message when there are identical rules, filters or conditions?
* Adjusted the UI layout, but instead of left aligning the elements, center-align seems more symmetric, how do you think about that? (I updated the screenshot at the top)

srijanreddy98 commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.html:
```diff
@@ -0,0 +1,270 @@
+<div *ngFor="let featureFlag of featureFlags; index as featureIndex;" class="container oppia-page-card oppia-long-text">
```
I don't see this `featureIndex` being used anywhere,  why is it present here?

MegrezZhu replied:
Yes it's no longer used, removed `featureIndex` here.

srijanreddy98 reviewed (approved):
LGTM! -- I have one minor nit though. PTAL!

MegrezZhu committed:
Fix style

MegrezZhu committed:
Fix test errors

BenHenning commented:
Code changes LGTM, but need to check the actual app again.

MegrezZhu committed:
Remove featureIndex

BenHenning commented:
New panel looks good! Thanks for making the changes.

BenHenning commented:
@MegrezZhu I think additional input sanitation would be a good idea--would it be a lot of work?

Also, will reserve approval until after #10185 is merged.

MegrezZhu commented:
> @MegrezZhu I think additional input sanitation would be a good idea--would it be a lot of work?
> 
> Also, will reserve approval until after #10185 is merged.

I can make a simple alert for these types of error, but it might take longer to design a nice UI (e.g. displaying errors just next to the related element).

MegrezZhu committed:
Add validation for identical rules/filters/conditions

MegrezZhu committed:
Fix linting errors

MegrezZhu committed:
Fix linting error

MegrezZhu committed:
Merge branch 'develop' into fe-admin-dev

MegrezZhu committed:
Merge branch 'fe-admin-dev' into fe-admin

MegrezZhu committed:
Merge branch 'develop' into fe-admin-dev

MegrezZhu committed:
Remove unused $apply

MegrezZhu commented:
Reopening as this PR is no longer blocked, @BenHenning PTAL!

MegrezZhu committed:
Merge branch 'develop' into fe-admin-dev

MegrezZhu committed:
Fix test errors

MegrezZhu committed:
Fix linting errors

BenHenning commented on core/templates/pages/admin-page/features-tab/admin-features-tab.component.html:
```diff
@@ -265,6 +263,57 @@ <h2 style="border-bottom: 1px solid #dee2e6;">{{featureFlag.name}}</h2>
     opacity: 1;
   }
+
+  .oppia-feature-tab-top-margined {
```
Perhaps instead: ``oppia-feature-top-margin`` or ``oppia-feature-top-with-margin``. How do we name similar classes elsewhere in the codebase?

Ditto for other 'margined' classes.

MegrezZhu replied:
Done.
In other files in the code base, there could be more fine grained control, like giving each html component a specific class name and assigning the margin properties to it, but since in this case most of the html elements only have margin properties so it would be simpler to use something like `top-margin` classname.

BenHenning replied:
Sounds good--thanks.

BenHenning reviewed (changes requested):
Thanks @MegrezZhu! PR still LGTM, but changing to 'request changes'. Other code owners should take a pass on this, and I plan to do a manual check of the panel once more tomorrow before approving.

BenHenning commented:
Ah, I just realized I'm the only one blocking the PR. I won't be able to check the functionality until tomorrow, but I'll prioritize this to unblock it.

MegrezZhu committed:
Refactor styling

BenHenning reviewed (approved):
Overall, @MegrezZhu I think this is fantastic! While playing around with the server, I actually created a new platform parameter, added it to gate backend code, then added it to gate frontend code & changed the filtering without restarting the server--for the most part, everything worked correctly. See my gist for a reference: https://gist.github.com/BenHenning/1d877317930da276991ca5ca9f305824/revisions. This was a great experience overall, though an end-to-end guide of how to add a new feature flag would be really helpful.

The only issue I ran into was when I tried having a disable rule with filter dev mode = server, and an enable rule with dev mode = debug and browser type = Chrome, but the feature didn't enable for me. Do you know what that is? It doesn't seem correct.

Other than the one issue above, everything seems good to me so approving this.

BenHenning commented:
Restarting the two failing workflows--hopefully they're just flaky.

MegrezZhu commented:
> Overall, @MegrezZhu I think this is fantastic! While playing around with the server, I actually created a new platform parameter, added it to gate backend code, then added it to gate frontend code & changed the filtering without restarting the server--for the most part, everything worked correctly. See my gist for a reference: https://gist.github.com/BenHenning/1d877317930da276991ca5ca9f305824/revisions. This was a great experience overall, though an end-to-end guide of how to add a new feature flag would be really helpful.
> 
> The only issue I ran into was when I tried having a disable rule with filter dev mode = server, and an enable rule with dev mode = debug and browser type = Chrome, but the feature didn't enable for me. Do you know what that is? It doesn't seem correct.
> 
> Other than the one issue above, everything seems good to me so approving this.

Thanks @BenHenning , by _a disable rule with filter dev mode = server, and an enable rule with dev mode = debug and browser type = Chrome_, do you mean the first rule is disabling with `server mode = dev` and the second is enabling with `server mode = dev & browser = Chrome`? (what is debug?).

The possible reason could be that
- the first rule is too broad so it's matched and applied before the second enabling rule, if so, we can move the broader rules to lower priority by moving them down
- have you re-login after changing the rules in the admin panel? the cached results can last for 12 hours

MegrezZhu committed:
Merge branch 'develop' into fe-admin-dev

BenHenning commented:
Thanks @MegrezZhu. Your interpretation is correct.

I created a new tab to verify that it was disabled--this also worked when verifying it was properly enabled.

I'm concerned that this will lead to confusion. Can we guarantee the order (e.g. either the first or last rule always wins in ambiguous cases)? If not, I think we should make that the case. The admin panel should also have some sort of indication of how the rules get evaluated (ideally, this would be a way to test a given configuration to see what the result would be, but short of that mentioning the ordering bit) might help reduce user error.

What do you think?

MegrezZhu committed:
Merge branch 'develop' into fe-admin-dev

MegrezZhu commented:
> Thanks @MegrezZhu. Your interpretation is correct.
> 
> I created a new tab to verify that it was disabled--this also worked when verifying it was properly enabled.
> 
> I'm concerned that this will lead to confusion. Can we guarantee the order (e.g. either the first or last rule always wins in ambiguous cases)? If not, I think we should make that the case. The admin panel should also have some sort of indication of how the rules get evaluated (ideally, this would be a way to test a given configuration to see what the result would be, but short of that mentioning the ordering bit) might help reduce user error.
> 
> What do you think?

@BenHenning It's guaranteed that the first rule always wins if its filters and conditions are matched -- it's been stated in the designed doc that "the value of the first matched rule is used", but we can emphasize this more explicitly in the document for administrators. Also, in order to reduce potential errors introduced by such confusions, in the future there can be some sort of "test tools" which can be used to test the status of features with given environment infos.

As for the CI error in the 'Travis CI - Pull Request' test, I believe it's irrelavent to this PR because I saw similar errors on the develop branch. It seems to have been resolved in the latest few commits, so I have just merged develop, let's see if the error will happen again.

Update: all tests passed.

BenHenning commented:
Thanks @MegrezZhu. I agree that we can add a "test the features" feature in the future (can you file a new issue to track this?), but in the meantime it seems too easy to mess this up without an explicit label. I think even just having a label in the UI that says "Note: the first rule that matches client criteria will be chosen" or similar would be helpful--we want to minimize admins specifically messing up gating criteria since that's exactly the situation that could result in accidentally releasing or not releasing functionality.

MegrezZhu committed:
Merge branch 'develop' into fe-admin-dev

MegrezZhu committed:
Add note to admin, fix style bug

MegrezZhu committed:
Merge branch 'develop' into fe-admin-dev

MegrezZhu commented:
> Thanks @MegrezZhu. I agree that we can add a "test the features" feature in the future (can you file a new issue to track this?), but in the meantime it seems too easy to mess this up without an explicit label. I think even just having a label in the UI that says "Note: the first rule that matches client criteria will be chosen" or similar would be helpful--we want to minimize admins specifically messing up gating criteria since that's exactly the situation that could result in accidentally releasing or not releasing functionality.

Added a boled note in the rules section on the UI:
![image](https://user-images.githubusercontent.com/21364971/95540663-5c7c8500-09bf-11eb-9cea-77205f175cd7.png)

Also I just created issue #10913 tracking the testing functionality.

DubeySandeep commented:
@MegrezZhu, Can you please resolve conflicts on this PR? (We can merge this PR once the conflicts are resolved!)

DubeySandeep commented:
@MegrezZhu, Any update on this PR?

MegrezZhu commented:
> @MegrezZhu, Can you please resolve conflicts on this PR? (We can merge this PR once the conflicts are resolved!)

@DubeySandeep I will resolve the conflicts shortly. Thanks for the notice.

MegrezZhu committed:
Merge branch 'develop' into fe-admin-dev

DubeySandeep commented:
@MegrezZhu, Lint checks are failing on this PR, PTAL!

MegrezZhu committed:
Fix indentation

BenHenning reviewed (approved):
Thanks @MegrezZhu, this looks great to me!

BenHenning commented:
Linter tests are now passing. E2e test failure seems unrelated, but need to wait for the others to finish before I can trigger a restart.

BenHenning commented:
Restarted 2 failing e2e test suites.

BenHenning commented:
@DubeySandeep do the backend tests usually take >2 hours to run?

DubeySandeep commented:
> @DubeySandeep do the backend tests usually take >2 hours to run?

@BenHenning, Yeah it takes around 2 hrs. 20 mins. to complete the backend test build.

BenHenning commented:
Thanks @DubeySandeep!

DubeySandeep commented:
Hi @MegrezZhu, it looks like the e2e test is failing with 500 errors after this PR is merged. Can you please confirm that it's happening because of this PR?

Link to fail build: https://app.circleci.com/pipelines/github/krishnarao22/oppia/15/workflows/991e9ef7-099a-4c56-8e32-e88fb2829db3/jobs/104

Error log:
```
F    ✗ should be correct for collaborators
      - Failed: Error while waiting for Protractor to sync with the page: "both angularJS testability and angular testability are undefined.  This could be either because this is a non-angular page or because your test involves client-side navigation, which can interfere with Protractor's bootstrapping.  See http://git.io/v4gXM for details"
      - Expected $.length = 1 to equal 0.
      Expected $[0] = Entry({ level: SEVERE, message: 'http://localhost:9001/signup?return_url=http%3A%2F%2Flocalhost%3A9001%2F - Failed to load resource: the server responded with a status of 500 (Internal Server Error)', timestamp: 1602676980501, type: '' }) to equal undefined.
ERROR    2020-10-14 12:03:00,891 wsgi.py:269] 
Traceback (most recent call last):
  File "/home/circleci/oppia_tools/google-cloud-sdk-304.0.0/google-cloud-sdk/platform/google_appengine/google/appengine/runtime/wsgi.py", line 240, in Handle
    handler = _config_handle.add_wsgi_middleware(self._LoadHandler())
  File "/home/circleci/oppia_tools/google-cloud-sdk-304.0.0/google-cloud-sdk/platform/google_appengine/google/appengine/runtime/wsgi.py", line 311, in _LoadHandler
    handler, path, err = LoadObject(self._handler)
  File "/home/circleci/oppia_tools/google-cloud-sdk-304.0.0/google-cloud-sdk/platform/google_appengine/google/appengine/runtime/wsgi.py", line 85, in LoadObject
    obj = __import__(path[0])
  File "/home/circleci/oppia/main.py", line 23, in <module>
    from core.controllers import acl_decorators
  File "/home/circleci/oppia/core/controllers/acl_decorators.py", line 25, in <module>
    from core.controllers import base
  File "/home/circleci/oppia/core/controllers/base.py", line 32, in <module>
    from core.domain import config_domain
  File "/home/circleci/oppia/core/domain/config_domain.py", line 23, in <module>
    from core.domain import caching_services
  File "/home/circleci/oppia/core/domain/caching_services.py", line 26, in <module>
    from core.domain import platform_parameter_domain
  File "/home/circleci/oppia/core/domain/platform_parameter_domain.py", line 42, in <module>
    ALLOWED_CLIENT_TYPES = constants.PLATFORM_PARAMETER_ALLOWED_CLIENT_TYPES
KeyError: 'PLATFORM_PARAMETER_ALLOWED_CLIENT_TYPES'
```

Also, wanted to confirm that there's isn't any merged PR made on top of these changes, right? (We may have to revert this PR to fix the issue and unblock other contributors quickly, so wanted to confirm.)

/cc @BenHenning

MegrezZhu commented:
@DubeySandeep thanks for letting me know. I just took a look at the e2e tests on the develop branch and several other PRs, it seems that they are working fine -- I didn't find these particular `KeyError: 'PLATFORM_PARAMETER_ALLOWED_CLIENT_TYPES'` errors in their logs.

If this is not the case or there are other PRs / tests that are also failing because of this PR, please let me know, otherwise this might be caused by that particular PR (I'm not sure though)

seanlip commented:
FYI -- this PR might have caused a flake in the frontend tests. See https://github.com/oppia/oppia/pull/10971/files#diff-fabe34989ce6e41037dad2ba29cd8de74de32aabcaea2bcbb8d7a0692d521c8bR369 for (what I hope is) a fix.

MegrezZhu commented:
> FYI -- this PR might have caused a flake in the frontend tests. See https://github.com/oppia/oppia/pull/10971/files#diff-fabe34989ce6e41037dad2ba29cd8de74de32aabcaea2bcbb8d7a0692d521c8bR369 for (what I hope is) a fix.

@seanlip Thanks!
